WIP Update Dependencies and Requirements to use `pyproject.toml`

mfixstsci opened the pull request:
Currently JWQL is using three different methods to track dependencies and build our python environment. It would be nice to shrink this down to one method and keep track of them in the `pyproject.toml`

mfixstsci committed:
Removing conda environments and pip requirements. Moving everything to pyproject.toml

mfixstsci committed:
Merge branch 'develop' into clean-dependency-lists

mfixstsci committed:
update postgres requirement in pyproject

zacharyburnett commented:
Looks good so far! You can likely remove `ruff` and `pytest`, and limit these dependencies only to those imported in the code. I'll draft up a release workflow too

zacharyburnett committed:
pare dependencies that aren't explicitly imported in code

mfixstsci committed:
Merge remote-tracking branch 'upstream/develop' into clean-dependency-lists

mfixstsci committed:
Removing pytest and ruff

mfixstsci committed:
Merge pull request #7 from zacharyburnett/pare_dependencies

pare dependencies that aren't explicitly imported in code

mfixstsci committed:
Adding upper and lower pins

mfixstsci commented:
@zacharyburnett Thanks for the PR, I went ahead and added the upper and lower bound pins. I also added the minimum python version. I guess the next steps would be to update the CI pipeline to get the tests passing. I am sure there is a nifty and much easier way to get the package installed with the pyproject.toml instead of executing a bunch of bash commands. Ideally the minimum version we would like to support at this point is 3.11 so the testing matrix will need to be updated to include python 3.11 and 3.12.

zacharyburnett committed:
test workflow

zacharyburnett committed:
build workflow

zacharyburnett committed:
python versions

mfixstsci committed:
Merge pull request #8 from zacharyburnett/ci_workflows

CI workflows

zacharyburnett committed:
add matrix to build workflow

zacharyburnett committed:
job name

mfixstsci committed:
Merge pull request #9 from zacharyburnett/build_matrix

add matrix to build workflow

mfixstsci committed:
updating python version for rtd

mfixstsci committed:
Merge remote-tracking branch 'origin/clean-dependency-lists' into clean-dependency-lists

mfixstsci committed:
setting strick python version for rtd

zacharyburnett committed:
use docs extra

zacharyburnett committed:
update docs dependencies

mfixstsci committed:
Merge pull request #10 from zacharyburnett/fix/rtd

fix ReadTheDocs build

mfixstsci commented:
@zacharyburnett so close, looks like something with the mast portal is causing the test failure. I will investigate further tomorrow.

zacharyburnett commented on .github/workflows/ci.yml:
```diff
@@ -1,54 +1,46 @@
 concurrency:
   group: ${{ github.workflow }}-${{ github.ref }}
   cancel-in-progress: true
 
```
```suggestion

defaults:
  run:
    shell: micromamba-shell {0}

```

zacharyburnett commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,43 @@
+    types: [ released ]
+  pull_request:
+  workflow_dispatch:
+
```
```suggestion

defaults:
  run:
    shell: micromamba-shell {0}

```

zacharyburnett commented on .github/workflows/ci.yml:
```diff
@@ -1,54 +1,46 @@
       - uses: actions/checkout@v4
-
-      - uses: mamba-org/provision-with-micromamba@v15
+      - uses: actions/setup-python@v5
```
```suggestion
      - uses: mamba-org/setup-micromamba@v1
```

zacharyburnett commented on .github/workflows/ci.yml:
```diff
@@ -1,54 +1,46 @@
-      - run: conda env export
-
-      - uses: supercharge/redis-github-action@1.4.0
+          python-version: ${{ matrix.python-version }}
+          cache: pip
+          cache-dependency-path: pyproject.toml
```
```suggestion
          environment-name: jwql-${{ runner.os }}-py${{ matrix.python-version }}
          environment-file: environment.yaml
          create-args: >-
            python=${{ matrix.python-version }}
          init-shell: none
          generate-run-shell: true
```

zacharyburnett commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,43 @@
+    name: freeze versions (Python ${{ matrix.python-version }})
+    steps:
+      - uses: actions/checkout@v4
+      - uses: actions/setup-python@v5
+        with:
+          python-version: ${{ matrix.python-version }}
```
```suggestion
      - uses: mamba-org/setup-micromamba@v1
         with:
          environment-name: jwql-${{ runner.os }}-py${{ matrix.python-version }}
          environment-file: environment.yaml
          create-args: >-
            python=${{ matrix.python-version }}
          init-shell: none
          generate-run-shell: true
```

zacharyburnett commented on .github/workflows/ci.yml:
```diff
@@ -1,54 +1,46 @@
-      max-parallel: 5
-      fail-fast: false
       matrix:
         os: [ubuntu-latest, macos-latest]
```
```suggestion
        os: [ "ubuntu-latest", "macos-latest" ]
```

zacharyburnett commented on .github/workflows/ci.yml:
```diff
@@ -1,54 +1,46 @@
         os: [ubuntu-latest, macos-latest]
-        python-version: [3.9, "3.10"]
-
+        python-version: ["3.11"]
```
```suggestion
        python-version: [ "3.11" ]
```

zacharyburnett reviewed (changes requested):
@mfixstsci use these changes with an `environment.yml` that includes `firefox`:
```yaml
# environment.yml

dependencies:
  - firefox
  - python
```

zacharyburnett commented:
superseded by #1568
